### ISSUE TITLE
NOT_VALIDATED always returns when specify "groupMemberOfAttribute“

### ISSUE THREAD

**Issue by wxwesther** [javaee/security-soteria#126]
```
   @LdapIdentityStoreDefinition(
	    url = "ldap://localhost:11389/",
	    callerBaseDn = "ou=caller,dc=securityapi,dc=net",
	    groupMemberOfAttribute = "notexisting"
)
```

group search is not configured in above LdapIdentityStoreDefinition, and below user exists. 

```
dn: uid=tom,ou=caller,dc=securityapi,dc=net
objectclass: top
objectclass: uidObject
objectclass: person
uid: tom
cn: Tom Rahman
sn: Tom
userPassword: secret1
```

NOT_VALIDATED always returned in result in below two cases:
 case1 ) groupMemberOfAttribute is specified as non-existing attribute("notexisting") 
 case2)  groupMemberOfAttribute is specified as existing attribute ("objectclass")

IMHO, case1)　should return VALID without any group．
　　　case2)  should return VALID with groups={top, uidObject, person}

**Comment by wmhopkins** [comment]
@wxwesther:

For **case1**, this is bad configuration. If we can detect the bad configuration, we will throw an exception. If not -- if all we can discover is that groups were not returned -- then we will return _VALID_, as you say, without groups (assuming that the validation step succeeded).

See my comments in #127 for an explanation of the status values and when an exception will be thrown.

For **case2**, I think the test is invalid. The `memberOf` functionality needs to be tested with a real `memberOf` attribute, or at least an attribute whose values are DNs, because the code must pull actual DN values (the DNs of the group objects) from the attribute, and then parse them to get the group names. That won't work if the test attribute doesn't contain DNs.

I'll look into what's going on with **case1**, and also fix the `LdapIdentityStore` to handle `memberOf` properly, and see if I can configure Unboundid or OpenDS to support `memberOf`.

**Comment by vinayvishal** [comment]
When `memberof` attribute is not available in person object, a `NullPointerException` is getting thrown.

```java
at org.glassfish.soteria.identitystores.LdapIdentityStore.retrieveGroupInformationMemberOf(LdapIdentityStore.java:270)
```
The problematic piece of code is following:

**LdapIdentityStore**

```java
Attribute memberOfAttribute = attributes.get(ldapIdentityStoreDefinition.groupMemberOfAttribute());
            memberOfValues = list(memberOfAttribute.getAll());
```

Since `memberOfAttribute` is null in this case, a null check is required here.

**from Javadoc**

> `javax.naming.directory.Attributes`
>
> `public abstract javax.naming.directory.Attribute get(String attrID)`
>                        
> Retrieves the attribute with the given attribute id from the attribute set.
>
> **Parameters:**
>
>`attrID` - The `non-null` id of the attribute to retrieve. If this attribute set ignores the character case of its attribute ids, the case of attrID is ignored.
>
>**Returns:**
>
>The attribute identified by `attrID`; **null if not found.**

I will soon provide a patch too for this.

In **case2** , I searched for documentation for both `openDS` and `Unboundid`, seems both of them don't support `memberof`.

* [Unboundid open source in memory database doesn't support memberof](https://stackoverflow.com/questions/42024558/unboundid-ldap-memberof-overlay)

Instead both seems to support virtual `isMemberOf` attribute, which gets generated dynamically.
* [isMemberOf in Unboundid](https://ping.force.com/Support/PingIdentityArticle?id=kA340000000PMxxCAG)
* [Is Member Of Virtual Attribute in OpenDS](https://docs.oracle.com/cd/E19476-01/821-0507/is-member-of-virtual-attribute.html)

**Comment by wmhopkins** [comment]
MemberOf support is fixed by PR #170 .

Note that the free version of the Unboundid in-memory server doesn't support dynamic generation of `memberOf` attributes for groups that the user is a member of. However, this behavior can be emulated for testing purposes by using the `seeAlso` attribute, which has DN syntax, and can be added to `person` objects and objects that extend `person`. See the `test.ldif` file and `LdapIdentityStoreDefinition` annotation used in the `test/app-ldap3` test in Soteria for an example of how to use `seeAlso`.

A similar emulation is almost certainly possible for the OpenDS directory. What's needed is a multi-valued attribute, defined by the OpenDS schema, that has DN syntax (extends `distinguishedName`) and can be added to user entries. I can help identify such an attribute if needed.
